modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h

```diff
@@ -0,0 +1,114 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "PorousFlowVariableBase.h"
+#include "PorousFlowFluidStateBase.h"
+
+class PorousFlowCapillaryPressure;
+
+/**
+ * Fluid state base class using a persistent set of primary variables for
+ * multiphase, single and multicomponent cases.
+ */
+template <bool is_ad>
+class PorousFlowFluidStateBaseMaterialTempl : public PorousFlowVariableBaseTempl<is_ad>
+{
+public:
+  static InputParameters validParams();
+
+  PorousFlowFluidStateBaseMaterialTempl(const InputParameters & parameters);
+
+protected:
+  virtual void initQpStatefulProperties() override;
+  virtual void computeQpProperties() override;
+
+  /// Size material property vectors and initialise with zeros
+  virtual void setMaterialVectorSize() const;
+
+  /**
+   * Calculates all required thermophysical properties and derivatives for each phase
+   * and fluid component
+   */
+  virtual void thermophysicalProperties() = 0;
+
+  /**
+   * Templated method that takes an ADReal value, and returns the
+   * raw value when is_ad = false, and the AD value when is_ad = true
+   * @param value AD value
+   * @return raw or AD value depending on is_ad
+   */
+  GenericReal<is_ad> genericValue(const ADReal & value);
```

Comment:
so weird that we dont have that elsewhere
@lindsayad do we not?

Reply:
if we dont, I could see this in MooseTypes.h.
what do you think?

Reply:
I would prefer not to. I generally don't like to add more APIs for dropping derivatives. But I would be open to changing my mind

Reply:
I'm fine with you having it in your module. And if other modules come up with justifiable uses then we could migrate it to the framework